registry/tests/team_payments_test.py

```diff
@@ -0,0 +1,197 @@
+
+        resp = self.app.get(
+            '/api/profile',
+            headers={
+                'Authorization': user1,
+            }
+        )
+        assert resp.status_code == requests.codes.ok
+
+        customer_create.assert_called_with(email=None, description='Team test_team')
+        subscription_create.assert_called_with(customer='cus_1', plan=PaymentPlan.TEAM_UNPAID.value)
+        customer_retrieve.assert_called_with('cus_1')
+
+        # Now access it as the second user, and verify that no customer gets created,
+        # and the info is the same as before.
+        customer_create.reset_mock()
+        subscription_create.reset_mock()
+
+        resp = self.app.get(
+            '/api/profile',
+            headers={
+                'Authorization': user2,
+            }
+        )
+        assert resp.status_code == requests.codes.ok
+
+        customer_create.assert_not_called()
+        subscription_create.assert_not_called()
+        customer_retrieve.assert_called_with('cus_1')
+
+    @mock_customer(plan=PaymentPlan.TEAM_UNPAID, have_credit_card=True)
+    def testUpgrade(self, customer):
+        admin = 'admin'
+        subscription = customer.subscriptions.data[0]
+        subscription.save.return_value = None
+
+        resp = self.app.post(
+            '/api/payments/update_plan',
+            data=dict(
+                plan=PaymentPlan.TEAM.value
+            ),
+            headers={
+                'Authorization': admin,
+            }
+        )
+        assert resp.status_code == requests.codes.ok
+        data = json.loads(resp.data.decode('utf8'))
+        assert data['plan'] == PaymentPlan.TEAM.value
+
+        assert subscription.plan == PaymentPlan.TEAM.value
+        subscription.save.assert_called_with()
+
+    @mock_customer(plan=PaymentPlan.TEAM, have_credit_card=True)
+    def testDowngrade(self, customer):
+        admin = 'admin'
+        subscription = customer.subscriptions.data[0]
+        subscription.save.return_value = None
+
+        resp = self.app.post(
+            '/api/payments/update_plan',
+            data=dict(
+                plan=PaymentPlan.TEAM_UNPAID.value
+            ),
+            headers={
+                'Authorization': admin,
+            }
+        )
+        assert resp.status_code == requests.codes.ok
+        data = json.loads(resp.data.decode('utf8'))
+        assert data['plan'] == PaymentPlan.TEAM_UNPAID.value
+
+        assert subscription.plan == PaymentPlan.TEAM_UNPAID.value
+        subscription.save.assert_called_with()
+
+    @mock_customer(plan=PaymentPlan.TEAM_UNPAID, have_credit_card=True)
+    def testUpdatePlanInvalid(self, customer):
+        admin = 'admin'
+        subscription = customer.subscriptions.data[0]
+        subscription.save.return_value = None
+
+        resp = self.app.post(
+            '/api/payments/update_plan',
+            data=dict(
+                plan=PaymentPlan.INDIVIDUAL.value
```

Comment:
this is the test case I was most looking for. And, to the point above, you could test the full cross-plan transitions easily if you had e.g. `TeamPlan` and `CloudPlan` in separate dicts.

Reply:
Well, `PaymentPlan` is a type, not a dict, so it really should contain all possible values. (Of course, it's Python, but in a strictly typed language that would be required.)

We could have sets like `TEAM_PLANS` and `CLOUD_PLANS` in addition to the type.

Reply:
subtypes?

Reply:
Can't even do that: https://docs.python.org/3/library/enum.html#restricted-subclassing-of-enumerations

Reply:
Actually, maybe... I'll play around with that